astropy/erfa/core.py

```diff
@@ -0,0 +1,326 @@
+    status codes that are warnings.
+    """
+
+    def generate_message(statcodes):
+        errcodes = np.unique(statcodes)
+
+        errcounts = dict((e, np.sum(statcodes==e)) for e in errcodes)
+
+        stat_msgs = STATUS_CODES[func_name]
+
+        elsemsg = STATUS_CODES[func_name].get('else')
+
+        if elsemsg is None:
+            default = 'Return code {0}'
+            errmsgs = dict((e, stat_msgs.get(str(e), default.format(e)))
+                           for e in errcodes)
+        else:
+            errmsgs = dict((e, stat_msgs.get(e, elsemsg)) for e in errcodes)
+
+        return ', '.join('{0} of "{1}"'.format(errcounts[e], errmsgs[e])
+                         for e in errcodes)
+
+    if np.any(statcodes < 0):
+        # errors present - only report the errors.
+        if statcodes.shape:
+            statcodes = statcodes[statcodes < 0]
+
+        err_msg = generate_message(statcodes)
+        raise ErfaError(
+            "ERFA function '{0}' yielded {1}".format(func_name, err_msg))
+
+    elif np.any(statcodes > 0):
+        # only warnings present
+        if statcodes.shape:
+            statcodes = statcodes[statcodes > 0]
+
+        warn_msg = generate_message(statcodes)
+        warnings.warn(
+            "ERFA function '{0}' yielded {1}".format(func_name, warn_msg),
+            ErfaWarning)
+
+
+def make_erfa_func(func_name, func_meta):
+    """
+    Given the (Python) function name for an ERFA function, and the function's
+    metadata as read out of the ``erfa.json`` file, returns a Python function
+    with the name and argument signature matching the related ERFA function
+    and that prepares the user's arguments for the vectorized implementation
+    of that function.
+    """
+
+    def erfa_func(*args):
+        return erfa_func_wrapper(func_name, func_meta, *args)
+
+    erfa_func.__doc__ = func_meta['doc']
+
+    func_args = tuple(arg['name'] for arg in func_meta['arguments']
+                      if arg['direction'] in ('in', 'inout'))
+
+    erfa_func = make_function_with_signature(erfa_func, args=func_args,
+                                             name=func_name)
+
+    # Add status codes to the STATUS_CODES mapping if this function returns a
+    # status
+    statuscodes = func_meta.get('statuscodes')
+    if statuscodes:
+        STATUS_CODES[func_name] = statuscodes
+
+    return erfa_func
+
+
+def erfa_func_wrapper(func_name, func_meta, *args):
+    """
+    A wrapper for the C/Cython vectorized versions of the ERFA functions.
+
+    This wrapper checks all input arguments and prepares input/output arrays
+    for the vectorized versions of the ERFA functions, then executes the
+    associated C loop.
+    """
+
+    iter_args = []
+
+    iter_shape = ()
+    arg_meta = func_meta['arguments']
+    make_outputs_scalar = NPYLT18
+
+    #Turn all inputs into arrays
+    for idx, (arg, meta) in enumerate(zip(args, arg_meta)):
+        arg = np.array(arg, dtype=CTYPE_TO_DTYPE[meta['ctype']],
```

Comment:
Actually, couldn't one do the copy for `inout` already here: just use `copy=arg_meta['direction'] == 'inout'`